Multiple cities support

johnjohndoe opened the pull request:
As already discussed with @torfsen I refactor the project to support displaying **market data of multiple cities**. City specific content is loaded from the respective GeoJSON file. The file name itself is constructed from the **URL hash**. If no hash is present, _Karlsruhe_ is hardcoded as the fallback.

I also updated the _README.md_ to reflect the changes.

johnjohndoe committed:
Enable loading different market files based on the URI hash.

johnjohndoe committed:
Update document title with the current city name.

johnjohndoe committed:
Read data source information from market file.

johnjohndoe committed:
Update legend headline with the current city name.

johnjohndoe committed:
Read initial map position information from market file.

torfsen commented:
Looks fine! Minor issues:
- If we want to include many cities it would be better to have a separate directory for the cities' JSON files. So I suggest to use `cities/karlsruhe.json` instead of `maerkte-karlsruhe.json`.
- Could you add data for another city? With only one city the mechanism is hard to test.
- The README should have a better title (currently it's just "Markets in ..."). How about "A Map of Farmer Markets" or something like that?

An idea regarding this feature (not necessarily for this PR): We could use a drop-down list of available cities in the title of the legend, i.e. something like "Wo ist Markt in [Karlsruhe▼]?". The styling should be unobtrusive, though.

johnjohndoe committed:
Move market data in cities folder.

johnjohndoe commented:
Thank you for your review.
- I moved the data file into a _cities_ folder.
- I updated the title in the README.
- I will send you the data file for Berlin via e-mail to allow testing. I prefer to add it in a separate pull request, though.

The navigation idea sounds reasonable. Let's move it into its own issue.

johnjohndoe commented:
:panda_face:

torfsen commented:
Looks good. Open issues:
- It's good to default to a fixed city for empty and invalid fragments. However, in these cases the result should be the same as if one had entered that city (Karlsruhe in our case) manually. So there should be no difference between `/index.html#karlsruhe`, `/index.html` and `/index.html#does-not-exist`. Hence, if the fragment is empty or invalid, then
  - Use the default city's legend title and data source (the latter seems to work already)
  - Use the default city's page title
  - Fix the URL fragment (`/index.html` and `/index.html#does-not-exist` become `/index.html/#karlsruhe`)
- The fragment should also be normalized if the city exists, i.e. `/index.html#KARLSruhe` should be changed to `/index.html#karlsruhe`. This makes sure that we have one canonical URL for each city.
- I would also prefer to store the default city's name directly (i.e. `DEFAULT_CITY = "karlsruhe"` instead of `DEFAULT_MARKET_FILE_PATH = "cities/karlsruhe.json"`). That makes the default independent from how and where the data is stored.

johnjohndoe commented:
Thanks for the review. :+1: I will look into this.

johnjohndoe commented:
:rooster: updated

johnjohndoe commented:
:walking: dumdidum

torfsen commented on js/main.js:
```diff
@@ -254,14 +261,120 @@ function initLegend() {
+ * otherwise the default city name;
+ */
+function getCityName() {
+    var parser = document.createElement('a');
```
Is there a reason why you create a separate DOM node for this task? Couldn't you use use [`window.location.hash`](https://stackoverflow.com/a/11662717/857390)?

johnjohndoe replied:
No reason. I refactored this.

torfsen commented:
Also: If I load the page, replace the fragment (`#karlsruhe`) by `#Karlsruhe` (upper-case `K`) and press enter then the content is loaded twice, with noticable flickering inbetween. Indeed, if I put a `console.log('init(..., "' + cityName + '")');` at the beginning of the `init` function then I get two equal debug outputs from that one reload: `init(..., "karlsruhe")` (both times with a lower-case `k`).

johnjohndoe committed:
Updates the URL hash in the browser.

johnjohndoe commented:
:hocho: The flickering and duplicate loading should be gone now.

torfsen commented:
Wonderful! Thanks for your endurance :stuck_out_tongue_winking_eye:
